File: ros2_ws/src/locomotion_controller/locomotion_controller/wireless_cmd_bridge.py
#!/usr/bin/env python3

import time

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSHistoryPolicy, QoSProfile, QoSReliabilityPolicy

from go2_msgs.msg import LocomotionCmd, PushEvent
from unitree_go.msg import WirelessController


def _clamp(value: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, value))


class WirelessCmdBridge(Node):
    _A_MASK = 1 << 8
    _Y_MASK = 1 << 11
    _DPAD_UP_MASK = 1 << 12
    _DPAD_RIGHT_MASK = 1 << 13
    _DPAD_DOWN_MASK = 1 << 14
    _DPAD_LEFT_MASK = 1 << 15

    def __init__(self) -> None:
        super().__init__("wireless_cmd_bridge")

        self.declare_parameter("wireless_topic", "/wirelesscontroller")
        self.declare_parameter("locomotion_cmd_topic", "/locomotion_cmd")
        self.declare_parameter("push_event_topic", "/data/push_event")
        self.declare_parameter("push_event_hz", 10.0)
        self.declare_parameter("publish_hz", 50.0)
        self.declare_parameter("cmd_timeout_s", 0.5)
        self.declare_parameter("deadzone", 0.05)

        # Axis scales for command generation.
        self.declare_parameter("scale_x", 0.6)     # ly -> x_vel
        self.declare_parameter("scale_y", -0.4)    # lx -> y_vel (flipped)
        self.declare_parameter("scale_yaw", -1.2)  # rx -> yaw_rate (flipped)

        # Keep fixed values for policy inputs not controlled by wireless sticks.
        self.declare_parameter("z_pos", 0.27)
        self.declare_parameter("gait_hz", 3.0)

        self.wireless_topic = str(self.get_parameter("wireless_topic").value)
        self.cmd_topic = str(self.get_parameter("locomotion_cmd_topic").value)
        self.push_event_topic = str(self.get_parameter("push_event_topic").value)
        self.push_event_hz = float(self.get_parameter("push_event_hz").value)
        self.publish_hz = float(self.get_parameter("publish_hz").value)
        self.cmd_timeout_s = float(self.get_parameter("cmd_timeout_s").value)
        self.deadzone = float(self.get_parameter("deadzone").value)
        self.scale_x = float(self.get_parameter("scale_x").value)
        self.scale_y = float(self.get_parameter("scale_y").value)
        self.scale_yaw = float(self.get_parameter("scale_yaw").value)
        self.z_pos = float(self.get_parameter("z_pos").value)
        self.gait_hz = float(self.get_parameter("gait_hz").value)

        qos = QoSProfile(
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10,
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
        )
        self.sub_wireless = self.create_subscription(
            WirelessController, self.wireless_topic, self.on_wireless, qos
        )
        self.pub_cmd = self.create_publisher(LocomotionCmd, self.cmd_topic, qos)
        self.pub_push_event = self.create_publisher(PushEvent, self.push_event_topic, qos)

        self._lx = 0.0
        self._ly = 0.0
        self._rx = 0.0
        self._latest_keys = 0
        self._last_wireless_time = None

        period = 1.0 / max(self.publish_hz, 1.0)
        self.timer = self.create_timer(period, self.on_timer)
        push_event_period = 1.0 / max(self.push_event_hz, 1.0)
        self.push_event_timer = self.create_timer(push_event_period, self.on_push_event_timer)
        self.get_logger().info(
            f"wireless_cmd_bridge running: {self.wireless_topic} -> {self.cmd_topic}, "
            f"push events -> {self.push_event_topic}"
        )

    def _apply_deadzone(self, v: float) -> float:
        if abs(v) < self.deadzone:
            return 0.0
        return v

    def on_wireless(self, msg: WirelessController) -> None:
        keys = int(msg.keys) & 0xFFFF
        self._latest_keys = keys

        self._lx = self._apply_deadzone(float(msg.lx))
        self._ly = self._apply_deadzone(float(msg.ly))
        self._rx = self._apply_deadzone(float(msg.rx))
        self._last_wireless_time = time.monotonic()

    def _label_from_keys(self, keys: int) -> int:
        if keys & self._Y_MASK:
            return int(PushEvent.UP)
        if keys & self._A_MASK:
            return int(PushEvent.DOWN)
        if keys & self._DPAD_UP_MASK:
            return int(PushEvent.FWD)
        if keys & self._DPAD_DOWN_MASK:
            return int(PushEvent.BACK)
        if keys & self._DPAD_LEFT_MASK:
            return int(PushEvent.LEFT)
        if keys & self._DPAD_RIGHT_MASK:
            return int(PushEvent.RIGHT)
        return int(PushEvent.NO_PUSH)

    def _publish_push_event(self, label: int) -> None:
        msg = PushEvent()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = "wireless_cmd_bridge"
        msg.label = int(label)
        self.pub_push_event.publish(msg)

    def on_push_event_timer(self) -> None:
        now = time.monotonic()
        stale = self._last_wireless_time is None or (now - self._last_wireless_time) > self.cmd_timeout_s
        if stale:
            self._publish_push_event(PushEvent.NO_PUSH)
            return
        self._publish_push_event(self._label_from_keys(self._latest_keys))

    def on_timer(self) -> None:
        now = time.monotonic()
        stale = self._last_wireless_time is None or (now - self._last_wireless_time) > self.cmd_timeout_s
        if stale:
            # Don't publish stale zeros continuously; let policy timeout handle it.
            return

        msg = LocomotionCmd()
        msg.stamp = self.get_clock().now().to_msg()
        msg.x_vel = _clamp(self._ly * self.scale_x, -abs(self.scale_x), abs(self.scale_x))
        msg.y_vel = _clamp(self._lx * self.scale_y, -abs(self.scale_y), abs(self.scale_y))
        msg.yaw_rate = _clamp(self._rx * self.scale_yaw, -abs(self.scale_yaw), abs(self.scale_yaw))
        msg.z_pos = self.z_pos
        msg.gait_hz = self.gait_hz
        self.pub_cmd.publish(msg)


def main() -> None:
    rclpy.init()
    node = WirelessCmdBridge()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/core/locomotion_controller_cpp/src/wireless_cmd_bridge.cpp
#include <chrono>
#include <cmath>
#include <functional>
#include <optional>

#include "go2_msgs/msg/locomotion_cmd.hpp"
#include "go2_msgs/msg/push_event.hpp"
#include "locomotion_controller_cpp/common.hpp"
#include "rclcpp/rclcpp.hpp"
#include "unitree_go/msg/wireless_controller.hpp"

namespace locomotion_controller_cpp
{

class WirelessCmdBridgeNode : public rclcpp::Node
{
public:
  WirelessCmdBridgeNode()
  : Node("wireless_cmd_bridge")
  {
    this->declare_parameter("push_event_hz", 10.0);
    this->declare_parameter("publish_hz", 50.0);
    this->declare_parameter("cmd_timeout_s", 0.5);
    this->declare_parameter("deadzone", 0.05);
    this->declare_parameter("scale_x", 0.6);
    this->declare_parameter("scale_y", -0.4);
    this->declare_parameter("scale_yaw", -1.2);
    this->declare_parameter("z_pos", 0.27);

    push_event_hz_ = this->get_parameter("push_event_hz").as_double();
    publish_hz_ = this->get_parameter("publish_hz").as_double();
    cmd_timeout_s_ = this->get_parameter("cmd_timeout_s").as_double();
    deadzone_ = this->get_parameter("deadzone").as_double();
    scale_x_ = this->get_parameter("scale_x").as_double();
    scale_y_ = this->get_parameter("scale_y").as_double();
    scale_yaw_ = this->get_parameter("scale_yaw").as_double();
    z_pos_ = this->get_parameter("z_pos").as_double();

    auto qos = rclcpp::QoS(rclcpp::KeepLast(10)).best_effort();
    sub_wireless_ = this->create_subscription<unitree_go::msg::WirelessController>(
      "/wirelesscontroller", qos,
      std::bind(&WirelessCmdBridgeNode::OnWireless, this, std::placeholders::_1));
    pub_cmd_ = this->create_publisher<go2_msgs::msg::LocomotionCmd>("/locomotion_cmd", qos);
    pub_push_event_ = this->create_publisher<go2_msgs::msg::PushEvent>("/data/push_event", qos);

    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(1.0 / std::max(1.0, publish_hz_)),
      std::bind(&WirelessCmdBridgeNode::OnTimer, this));
    push_event_timer_ = this->create_wall_timer(
      std::chrono::duration<double>(1.0 / std::max(1.0, push_event_hz_)),
      std::bind(&WirelessCmdBridgeNode::OnPushEventTimer, this));

    RCLCPP_INFO(
      this->get_logger(),
      "wireless_cmd_bridge running: /wirelesscontroller -> /locomotion_cmd, push events -> /data/push_event");
  }

private:
  static constexpr int kAMask = 1 << 8;
  static constexpr int kYMask = 1 << 11;
  static constexpr int kDpadUpMask = 1 << 12;
  static constexpr int kDpadRightMask = 1 << 13;
  static constexpr int kDpadDownMask = 1 << 14;
  static constexpr int kDpadLeftMask = 1 << 15;

  double ApplyDeadzone(const double value) const
  {
    return std::abs(value) < deadzone_ ? 0.0 : value;
  }

  bool IsStale() const
  {
    if (!last_wireless_time_.has_value())
    {
      return true;
    }
    const auto age_s =
      std::chrono::duration<double>(std::chrono::steady_clock::now() - *last_wireless_time_).count();
    return age_s > cmd_timeout_s_;
  }

  std::uint8_t LabelFromKeys(const int keys) const
  {
    if ((keys & kYMask) != 0)
    {
      return go2_msgs::msg::PushEvent::UP;
    }
    if ((keys & kAMask) != 0)
    {
      return go2_msgs::msg::PushEvent::DOWN;
    }
    if ((keys & kDpadUpMask) != 0)
    {
      return go2_msgs::msg::PushEvent::FWD;
    }
    if ((keys & kDpadDownMask) != 0)
    {
      return go2_msgs::msg::PushEvent::BACK;
    }
    if ((keys & kDpadLeftMask) != 0)
    {
      return go2_msgs::msg::PushEvent::LEFT;
    }
    if ((keys & kDpadRightMask) != 0)
    {
      return go2_msgs::msg::PushEvent::RIGHT;
    }
    return go2_msgs::msg::PushEvent::NO_PUSH;
  }

  void PublishPushEvent(const std::uint8_t label)
  {
    go2_msgs::msg::PushEvent msg;
    msg.header.stamp = this->now();
    msg.header.frame_id = "wireless_cmd_bridge";
    msg.label = label;
    pub_push_event_->publish(msg);
  }

  void OnWireless(const unitree_go::msg::WirelessController::SharedPtr msg)
  {
    latest_keys_ = static_cast<int>(msg->keys) & 0xFFFF;
    lx_ = ApplyDeadzone(static_cast<double>(msg->lx));
    ly_ = ApplyDeadzone(static_cast<double>(msg->ly));
    rx_ = ApplyDeadzone(static_cast<double>(msg->rx));
    last_wireless_time_ = std::chrono::steady_clock::now();
  }

  void OnPushEventTimer()
  {
    if (IsStale())
    {
      PublishPushEvent(go2_msgs::msg::PushEvent::NO_PUSH);
      return;
    }
    PublishPushEvent(LabelFromKeys(latest_keys_));
  }

  void OnTimer()
  {
    if (IsStale())
    {
      return;
    }

    go2_msgs::msg::LocomotionCmd msg;
    msg.stamp = this->now();
    msg.x_vel = Clamp(ly_ * scale_x_, -std::abs(scale_x_), std::abs(scale_x_));
    msg.y_vel = Clamp(lx_ * scale_y_, -std::abs(scale_y_), std::abs(scale_y_));
    msg.yaw_rate = Clamp(rx_ * scale_yaw_, -std::abs(scale_yaw_), std::abs(scale_yaw_));
    msg.z_pos = z_pos_;
    pub_cmd_->publish(msg);
  }

  rclcpp::Subscription<unitree_go::msg::WirelessController>::SharedPtr sub_wireless_;
  rclcpp::Publisher<go2_msgs::msg::LocomotionCmd>::SharedPtr pub_cmd_;
  rclcpp::Publisher<go2_msgs::msg::PushEvent>::SharedPtr pub_push_event_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::TimerBase::SharedPtr push_event_timer_;

  std::optional<std::chrono::steady_clock::time_point> last_wireless_time_;
  double lx_{0.0};
  double ly_{0.0};
  double rx_{0.0};
  int latest_keys_{0};

  double push_event_hz_{10.0};
  double publish_hz_{50.0};
  double cmd_timeout_s_{0.5};
  double deadzone_{0.05};
  double scale_x_{0.6};
  double scale_y_{-0.4};
  double scale_yaw_{-1.2};
  double z_pos_{0.27};
};

}  // namespace locomotion_controller_cpp

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<locomotion_controller_cpp::WirelessCmdBridgeNode>());
  if (rclcpp::ok())
  {
    rclcpp::shutdown();
  }
  return 0;
}
